site/learn/expert-advice/security/breached-password-detection.md

```diff
@@ -0,0 +1,155 @@
+
+However, for a quick solution, start with an API integration, perhaps using a [library such as these](https://haveibeenpwned.com/API/Consumers). It's certainly better to use an API integration than to not build this feature. Later, you can build your own data ingestion system as your use increases or you need more control. 
+
+In any event, you'll have a source of compromised credentials and a way to check to see if a password is in that set.
+
+## Configuring your authentication system to check for leaked passwords
+
+Now you'll need to hook into your user management service. Depending on what kind of auth system you have, the integration may be more or less difficult. However, you'll want to check passwords at a couple of different times during the lifecycle of a user's interactions with your systems:
+
+* When they register or an account is created for them.
+* When a user changes a password.
+* When a user logs in.
+* Periodically checking against all users.
+
+You should enable checks on the first two events to prevent known compromised credentials from entering your system. The registration or password change should fail, with a clear message, if the user provides a publicly known password. You will want to enforce this even when a customer service or administrative user creates an account.
+
+### Check even when the password isn't changing
+
+The last two deserve a bit of explanation. Suppose a user signs up on Example.com with a great password. Then they come to your site and sign up with the same great password. They continue to use your site for months, but forget about Example.com. 
+
+Then, Example.com is breached. They may send out a notice, but your user may not receive it or may not change their password. This [study from 2020 (PDF)](https://www.ieee-security.org/TC/SPW2020/ConPro/papers/bhagavatula-conpro20.pdf) covers a small dataset, approximately 250 users over two years. It found that of the 60ish users who had accounts on breached domains, only 13% changed their password within three months of the breach announcement.
+
+If you only check for compromise when the password is created at registration or modified by the end user, you'll end up with users who have credentials that have been compromised by breaches external to your system after account creation. 
+
+The Example.com breach affected your system through the vector of the reused password. Check for breached passwords whenever a user logs in and set up a scheduled job to check credentials of users who haven't logged in.
+
+### What does a breached password detection look like?
+
+The actual implementation of the password check depends on how you are receiving the compromised dataset, but often you have a list of hashed passwords with a certain algorithm (SHA-1 or SHA-256, for example). When the user enters their credential, hash that using the same algorithm and then see if that value is present in the datasets. If it is, then the password has been compromised. 
```

Comment:
Most databases are actually plaintext. Databases with hashed passwords are only good if they contain weak hashes and no load-factor (iterations). The attack vector is to get the plaintext password or generate another string that collides with the hash (really hard to do) since you can't type a hash into a password field. Therefore, from what we know, plaintext datasets are really the only ones that should be considered.